### ISSUE TITLE
boolean spgemm: can I avoid numeric phase?

### ISSUE THREAD

**Issue by cjain7** [kokkos/kokkos-kernels#422]
I'm using spgemm for an application in biological data analysis. In my application, I've boolean matrices, and I'm only interested in boolean matrix multiplication. From the output, I only need the non-zero pattern, i.e., the column entries and the row-map, but I don't care about values (they all will be >=1). I'm using kokkos-kernels as an external library, and calling spgemm routines using the [API recommended by you](https://github.com/kokkos/kokkos-kernels/wiki/SPARSE-3%3A%3Aspgemm).

I read your [spgemm paper](http://www.cs.sandia.gov/~srajama/publications/SPGEMM-ASHES17.pdf) this week, where you describe the role of symbolic and numeric phases. My question is following: As I don't care about output values, **can I avoid the numeric phase, and just use symbolic phase?** I ask this because majority of time (>90%) is being consumed in the numeric phase in my benchmarks. If possible, please let me know a convenient way to go about it.

The current output (using the API) of symbolic phase is row-map, but if I could also get the column indices (i.e., entries), that would be fantastic. (I am using this library with openmp support on Intel CPUs)

**Comment by srajama1** [comment]
We do have the column indices info in symbolic phase, but don't store it. We just store the column map. It will be a small change. Will you be interested in changing it and contributing back ?

**Comment by cjain7** [comment]
Thanks for responding. I too realized this when I read through the code to have a better understanding of the symbolic phase (both uncompressed and compressed modes). 

Storing column indices would make symbolic phase more expensive I assume (as we don't have row-map in the beginning), but unclear to me whether we're sure to get any advantage over the current version that uses symbolic + numeric. It would be good to know your thoughts about this...

**Comment by srajama1** [comment]
Yes, it is a chicken and egg thing. Having the rowmap made life simpler, so we did it this way. There could be some advantage in finding this out during symbolic, but it would be small. For example, we do one phase for triangle counting, but we have an additional filtering in triangle counting in addition to multiply. This results in advantage that was worth it do one phase.

**Comment by Char-Aznable** [comment]
I'm interested in using spmv snd spgemm for binary matrix as well. It seems to me the right thing to do would be to specialize crsmatrix for bool to completely avoid storing the values of one and overload the kernel for it.
